core/src/main/java/aima/core/search/basic/uninformed/DepthFirstSearch.java

```diff
@@ -0,0 +1,126 @@
+import aima.core.search.api.Problem;
+import aima.core.search.api.SearchController;
+import aima.core.search.api.SearchForActionsFunction;
+import aima.core.search.basic.support.BasicNodeFactory;
+import aima.core.search.basic.support.BasicSearchController;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * <pre>
+ * function DEPTH-FIRST-SEARCH(problem) returns a solution, or failure
+ *   node &larr; a node with STATE = problem.INITIAL-STATE, PATH-COST=0
+ *   if problem.GOAL-TEST(node.STATE) then return SOLUTION(node)
+ *   frontier &larr; a LIFO queue with node as the only element
+ *   explored &larr; an empty set
+ *   loop do
+ *       if EMPTY?(frontier) then return failure
+ *       node &larr; POP(frontier) // chooses the deepest node in the frontier
+ *       add node.STATE to explored
+ *       for each action in problem.ACTIONS(node.STATE) do
+ *           child &larr; CHILD-NODE(problem, node, action)
+ *           if child.STATE is not in explored or frontier then
+ *               if problem.GOAL-TEST(child.STATE) then return SOLUTION(child)
+ *               frontier &larr; INSERT(child, frontier)
+ * </pre>
+ *
+ * <b>Figure ??</b> Depth-first graph search.
+ *
+ * <p>The depth-first search algorithm is an instance of the graph-search algorithm;
+ * whereas breadth-first search uses a FIFO queue, depth-first-search uses a LIFO
+ * queue which means that the most recently generated node is chosen for expansion.
+ * This must be the deepest unexpanded node when it was selected. The graph-search
+ * version avoids repeated states and redundant paths and is complete in finite
+ * state spaces because it will eventually expand every node. There is one slight
+ * tweak on the general graph-search algorithm, which is that the goal test is
+ * applied to each node when it is generated rather than when it is selected for
+ * expansion.
+ */
+public class DepthFirstSearch<A, S> implements SearchForActionsFunction<A, S> {
+  // function DEPTH-FIRST-SEARCH(problem) returns a solution, or failure
+  @Override
+  public List<A> apply(Problem<A, S> problem) {
+    // node <- a node with STATE = problem.INITIAL-STATE, PATH-COST=0
+    Node<A, S> node = newRootNode(problem.initialState(), 0);
+    // if problem.GOAL-TEST(node.STATE) then return SOLUTION(node)
+    if (isGoalState(node, problem)) {
+      return solution(node);
+    }
+    // frontier <- a LIFO queue with node as the only element
+    Queue<Node<A, S>> frontier = newLIFOQueue(node);
+    // explored <- an empty set
+    Set<S> explored = newExploredSet();
+    // loop do
+    while (true) {
+      // if EMPTY?(frontier) then return failure
+      if (frontier.isEmpty()) {
+        return failure();
+      }
+      // node <- POP(frontier) // chooses the deepest unexplored node in the frontier
+      node = frontier.remove();
+      // add node.STATE to explored
```

Comment:
All these comments are not necessary, because they just duplicate the code

Reply:
I agree, the comments were added just to remain consistent with the already implemented algorithms. I feel that the project currently requires a set of contributing guidelines in order maintain and welcome consistent contributions in the future.

Reply:
@utsavoza I am working on the guidelines. I will add them after completion and discussion with the maintainers. Besides, you are welcome to give suggestions.